File: packages/sdks/server-sdks/gatrix-dotnet-server-sdk/src/Gatrix.Server.Sdk/Evaluation/Strategies/GradualRolloutSessionIdStrategy.cs
namespace Gatrix.Server.Sdk.Evaluation.Strategies;

/// <summary>
/// Gradual rollout based on session ID hash.
/// </summary>
public class GradualRolloutSessionIdStrategy : IStrategy
{
    public string Name => "gradualRolloutSessionId";

    public bool IsEnabled(Models.StrategyParameters? parameters, Models.EvaluationContext context, string flagName)
    {
        var pct = parameters?.Percentage ?? 0;
        if (pct >= 100) return true;
        if (pct <= 0) return false;
        if (string.IsNullOrEmpty(context.SessionId)) return false;

        var groupId = parameters?.GroupId ?? flagName;
        var percentage = StrategyUtils.CalculatePercentage(context.SessionId, groupId);
        return percentage <= pct;
    }
}

File: packages/evaluator/src/strategies/gradual-rollout-session-id-strategy.ts
/**
 * Gradual Rollout Session ID Strategy
 *
 * Percentage-based rollout using sessionId hash.
 * Stickiness is fixed to sessionId (not configurable).
 */

import {
  EvaluationContext,
  StrategyParameters,
  StrategyEvaluationResult,
} from '@gatrix/shared';
import { Strategy } from './strategy';
import { normalizedStrategyValue } from './util';

export class GradualRolloutSessionIdStrategy extends Strategy {
  constructor() {
    super('gradualRolloutSessionId');
  }

  isEnabled(
    parameters: StrategyParameters,
    context: EvaluationContext
  ): boolean {
    const { sessionId } = context;
    if (!sessionId) return false;
    const percentage = Number(parameters.percentage ?? 0);
    const groupId = parameters.groupId || '';
    const enabled =
      percentage > 0 &&
      normalizedStrategyValue(sessionId, groupId) <= percentage;
    return enabled;
  }

  isEnabledWithDetails(
    parameters: StrategyParameters,
    context: EvaluationContext
  ): StrategyEvaluationResult {
    const { sessionId } = context;
    if (!sessionId) {
      return { enabled: false, reason: 'No sessionId in context' };
    }

    const percentage = Number(parameters.percentage ?? 0);
    const groupId = parameters.groupId || '';
    const normalizedValue = normalizedStrategyValue(sessionId, groupId);
    const enabled = percentage > 0 && normalizedValue <= percentage;

    return {
      enabled,
      reason: enabled
        ? `Percentage ${percentage}%: normalized(sessionId="${sessionId}") = ${normalizedValue} <= ${percentage}`
        : `Percentage ${percentage}%: normalized(sessionId="${sessionId}") = ${normalizedValue} > ${percentage}`,
      details: { sessionId, percentage, groupId, normalizedValue },
    };
  }
}
